src/ooni/ooni_test_impl.hpp

```diff
@@ -135,118 +138,61 @@ class OoniTestImpl : public mk::NetTest, public mk::NonCopyable,
-
     std::string probe_ip = "127.0.0.1";
     std::string probe_asn = "AS0";
     std::string probe_cc = "ZZ";
+    bool needs_input = false;
 
     OoniTestImpl() : OoniTestImpl("", Settings()) {}
 
-    virtual ~OoniTestImpl() {
-        delete input;
-        input = nullptr;
-    }
+    virtual ~OoniTestImpl() {}
 
     OoniTestImpl(std::string input_filepath_)
         : OoniTestImpl(input_filepath_, Settings()) {}
 
     OoniTestImpl(std::string input_filepath_, Settings options_)
         : NetTest(input_filepath_, options_),
-        test_name("net_test"), test_version("0.0.1") {
-            mk::utc_time_now(&test_start_time);
-        }
-
-    std::string get_output_file_path() {
-        if (output_filepath == "") {
-            output_filepath = generate_output_file_path();
-        }
-        return output_filepath;
-    }
-
-    InputGenerator *input_generator() {
-        return new InputFileGenerator(input_filepath, logger);
-    }
+        test_name("net_test"), test_version("0.0.1") {}
 
     void begin(Callback<> cb) override {
+        mk::utc_time_now(&test_start_time);
         geoip_lookup([=]() {
             open_report();
-            if (input_filepath != "") {
-                logger->debug("net_test: found input file");
-                if (input != nullptr) {
-                    delete input;
-                    input = nullptr;
+            if (needs_input) {
+                if (input_filepath == "") {
+                    // TODO: allow to pass error to cb()
+                    logger->warn("an input file is required");
+                    cb();
+                    return;
                 }
-                input = input_generator();
-                run_next_measurement(std::move(cb));
-            } else {
-                logger->debug("net_test: no input file");
-                struct tm measurement_start_time;
-                double start_time;
-                mk::utc_time_now(&measurement_start_time);
-                start_time = mk::time_now();
-                logger->debug("net_test: calling setup");
-                setup();
-                main(options, [=](report::Entry test_keys) {
-                    report::Entry entry;
-                    entry["test_keys"] = test_keys;
-                    entry["input"] = "";
-                    entry["measurement_start_time"] = *mk::timestamp(&measurement_start_time);
-                    entry["test_runtime"] = mk::time_now() - start_time;
-                    logger->debug("net_test: tearing down");
-                    teardown();
-                    file_report.write_entry(entry);
-                    logger->debug("net_test: written entry");
-                    logger->debug("net_test: reached end of input");
+                input_generator.reset(new std::ifstream(input_filepath));
+                if (!input_generator->good()) {
+                    logger->warn("cannot read input file");
                     cb();
```

Comment:
Here we basically have the checks on the input file inline. It used to raise exceptions but that is not so much async. I'd like the final callback to receive an error when we can tell the users things such as missing file.